compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -646,4 +647,22 @@ object TreeChecker {
       tp
     }
   }.apply(tp0)
+
+  /** Check that the tree only contains legal children trees */
+  object TreeNodeChecker extends untpd.TreeTraverser:
+    import untpd._
+    def run(tree: Tree)(using Context) = if !tree.isInstanceOf[TypeTree] then traverse(tree)
+    def traverse(tree: Tree)(using Context) = tree match
+      case t: TypeTree                             => assert(assertion = false, t)
+      case t @ TypeApply(fun, _targs)              => traverse(fun)
+      case t @ New(_tpt)                           =>
+      case t @ Typed(expr, _tpt)                   => traverse(expr)
+      case t @ Closure(env, meth, _tpt)            => traverse(env); traverse(meth)
+      case t @ SeqLiteral(elems, _elemtpt)         => traverse(elems)
+      case t @ ValDef(_, _tpt, _)                  => traverse(t.rhs)
+      case t @ DefDef(_, paramss, _tpt, _)         => for params <- paramss do traverse(params); traverse(t.rhs)
+      case t @ TypeDef(_, _rhs)                    =>
+      case t @ Template(constr, _parents, self, _) => traverse(constr); traverse(self); traverse(t.body)
```

Comment:
Parents should also be traversed as we could have terms in them. Consider
```
class C extends B(Int: Int)
```

where both `Int`s are `TypeTree`s